### PR TITLE
 [Qt] Warn if fallback fee has been used

### PR DESCRIPTION
**Pull request by jonasschnelli**
This PR appends a warning at the send coins confirmation text in case the fallback fee was used.

bildschirmfoto 2017-12-13 um 13 13 02

### CODE REVIEW

File: src/qt/sendcoinsdialog.cpp [hunk 1]
```diff
@@ -346,6 +346,13 @@ void SendCoinsDialog::on_sendButton_clicked()
         questionString.append("</span>");
     }
 
+    if (prepareStatus.m_fallback_fee_used) {
+        questionString.append("<hr /><span style='color:#aa0000;'>");
+        questionString.append(tr("WARNING")+":</span> <span>");
+        questionString.append(tr("A static fallback transaction fee has been used due to missing fee estimations. Please consider waiting a few blocks until fee estimation is ready. Your transaction will very likely get stuck or you overpay in fees."));
```
Comment (promag) [inline src/qt/sendcoinsdialog.cpp]: s/estimations/estimation.

Comment (promag) [reply src/qt/sendcoinsdialog.cpp]: I would reorder the sentenses: 1, 2, 3 -> 3, 1, 2. Also, avoid you/your?

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -961,7 +961,7 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
      * @note passing nChangePosInOut as -1 will result in setting a random position
      */
     bool CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletTx& wtxNew, CReserveKey& reservekey, CAmount& nFeeRet, int& nChangePosInOut,
-                           std::string& strFailReason, const CCoinControl& coin_control, bool sign = true);
+                           std::string& strFailReason, const CCoinControl& coin_control, FeeCalculation& feeCalc, bool sign = true);
```
Comment (promag) [inline src/wallet/wallet.h]: fee_calc? This line already uses new convention. Same for others.

File: src/qt/walletmodel.h [hunk 3]
```diff
@@ -144,13 +144,15 @@ class WalletModel : public QObject
...
+        SendCoinsReturn(StatusCode _status = OK, QString _reasonCommitFailed = "", bool _m_fallback_fee_used = false)
             : status(_status),
-              reasonCommitFailed(_reasonCommitFailed)
+              reasonCommitFailed(_reasonCommitFailed),
+              m_fallback_fee_used(_m_fallback_fee_used)
         {
         }
         StatusCode status;
         QString reasonCommitFailed;
+        bool m_fallback_fee_used;
```
Comment (promag) [inline src/qt/walletmodel.h]: These should be const?

File: src/qt/walletmodel.h [hunk 4]
```diff
@@ -144,13 +144,15 @@ class WalletModel : public QObject
     // Return status record for SendCoins, contains error id + information
     struct SendCoinsReturn
     {
-        SendCoinsReturn(StatusCode _status = OK, QString _reasonCommitFailed = "")
+        SendCoinsReturn(StatusCode _status = OK, QString _reasonCommitFailed = "", bool _m_fallback_fee_used = false)
```
Comment (promag) [inline src/qt/walletmodel.h]: Drop `_m_` prefix? `, bool fallback_fee_used = false)`.

File: src/qt/walletmodel.cpp [hunk 5]
```diff
@@ -299,7 +300,7 @@ WalletModel::SendCoinsReturn WalletModel::prepareTransaction(WalletModelTransact
             return AbsurdFee;
     }
 
-    return SendCoinsReturn(OK);
+    return SendCoinsReturn(OK, "", (feeCalc.reason == FeeReason::FALLBACK));
```
Comment (promag) [inline src/qt/walletmodel.cpp]: Drop `()`?

File: src/qt/walletmodel.cpp [hunk 6]
```diff
@@ -285,7 +286,7 @@ WalletModel::SendCoinsReturn WalletModel::prepareTransaction(WalletModelTransact
         {
             if(!fSubtractFeeFromAmount && (total + nFeeRequired) > nBalance)
             {
-                return SendCoinsReturn(AmountWithFeeExceedsBalance);
+                return SendCoinsReturn(AmountWithFeeExceedsBalance, "", (feeCalc.reason == FeeReason::FALLBACK));
```
Comment (promag) [inline src/qt/walletmodel.cpp]: Drop `()`?

**Review by promag — COMMENTED** [review]
utACK 28a47e7. Some nits though.

File: src/qt/sendcoinsdialog.cpp (hunk 1 above)
Comment (jonasschnelli) [reply src/qt/sendcoinsdialog.cpp]: I think the order is correct (cause, fix, possible issue).

File: src/qt/walletmodel.h (hunk 3 above)
Comment (jonasschnelli) [reply src/qt/walletmodel.h]: I keep it non const for now to avoid changing code that receives that struct, also, other variables are also non-const.

**Comment by jonasschnelli** [comment]
Fixed @promag points

**Comment by laanwj** [comment]
Concept ACK.

**Comment by morcos** [comment]
Another option for the GUI would be to just prevent using a fallback fee at all.  We already have the option to manually set the fee.  We could instead refuse to send the transaction unless the user manually sets the fee. I think I would prefer that approach.

Actually it might not be terrible to do a similar thing for RPC.  Where any sendto call fails if fee estimation is needed and doesn't work with an error message telling you you must specify the fee (hopefully the option to specify the fee exists for all RPC sending calls).  But at the very least we could do it for QT.

**Comment by jonasschnelli** [comment]
@morcos see #11882.
My idea was to follow these steps:
1. Allow to disable fallbackfee (but off by default) (#11882)
2. Force RBF BIP125 if fallbackfee is active (#11882)
3. Warn in the GUI if fallbackfee is present (this PR)
4. Disable fallback fee by default in GUI (upcoming if #11882 gets merged)
5. Disable fallback fee by default in RPC (upcoming if #11882 gets merged, but later)

**Comment by morcos** [comment]
@jonasschnelli commented there.
too many options.  lets just be more forceful about defaulting to the "right" thing.  and if people want something different they have a way of manually doing that.

**Comment by jonasschnelli** [comment]
@morcos. Yes. Sounds good.
This PR though is completely disconnected from the default value,... lets continue the default value in #11882.

**Comment by morcos** [comment]
Sure you're right

**Comment by thijstriemstra** [comment]
"static fallback transaction fee" does not mean a lot to me (and most people). Can this be rephrased to something human readable?

**Comment by MarcoFalke** [comment]
Is this still needed, since the user explicitly set the fallbackfee (and thus is aware of it)?

I am assuming #11882.

**Comment by thijstriemstra** [comment]
[I don't want to block progress on any PRs but] I'd have to google this 'static fallback transaction fee' before i'd feel happy to make the transaction when reading such a warning..

**Comment by sipa** [comment]
@thijstriemstra Would it help to put the "Please consider waiting a few blocks until fee estimation is ready" first?

**Comment by thijstriemstra** [comment]
In a way, but 'a few', can we be more specific? or less detailed? :) as long as cryptic statements like 'static fallback' are avoided I think. Thanks for the feedback!

**Comment by laanwj** [comment]
I think this can be closed now that the fallback fee has been disabled on mainnet (#11882).

**Pull request closed without merging**